### PR TITLE
Add 3D convolution transposed operator.

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/5116

### CODE REVIEW

**Comment by zchen0211** [comment]
Generally looks good. Is this the C3D model as in http://vlg.cs.dartmouth.edu/c3d/?

**Comment by chengduoZH** [comment]
> Is this the C3D model as in http://vlg.cs.dartmouth.edu/c3d/?

Yes. Conv3D and DeConv3D are needed in video processing or medical image processing. This requirement was put forward by IDL medical image processing team.

**Comment by zchen0211** [comment]
I see. Just had a look and still looks pretty good to me.

**Comment by zchen0211** [comment]
Approved from my side. Maybe Qingqing and Daoyuan could have a double check :)

File: paddle/operators/conv_transpose_op.h [hunk 1]
```diff
@@ -250,5 +257,210 @@ class GemmConv2DTransposeGradKernel : public framework::OpKernel<T> {
...
+    const int w = input->dims()[4];
+
+    const int k_d = filter.dims()[2];
+    const int k_h = filter.dims()[3];
+    const int k_w = filter.dims()[4];
+
+    const int c = output->dims()[1];  // output channels
+    const int o_d = output->dims()[2];
+    const int o_h = output->dims()[3];
+    const int o_w = output->dims()[4];
```
Comment (qingqing01) [inline paddle/operators/conv_transpose_op.h]: `DDim`里存的是`int64_t`，如果编译有warning的话，可以用`int64_t`类型。

File: paddle/operators/conv_transpose_op.h [hunk 2]
```diff
@@ -250,5 +257,210 @@ class GemmConv2DTransposeGradKernel : public framework::OpKernel<T> {
...
+      for (int i = 0; i < batch_size; ++i) {
+        // batch with size (c, o_d, o_h, o_w)
+        Tensor output_grad_batch =
+            output_grad->Slice(i, i + 1).Resize(output_shape);
+        // input batch
+        Tensor in_batch = input->Slice(i, i + 1).Resize(input_matrix_shape);
+
+        // vol2col: (c * d * h * w, k_d * k_h * k_w)
+        vol2col(context.device_context(), output_grad_batch, col, strides[0],
+                strides[1], strides[2], paddings[0], paddings[1], paddings[2]);
```
Comment (qingqing01) [inline paddle/operators/conv_transpose_op.h]: line 396 和 line 431的if分支分开写，导致调用了两次 vol2col，这样多了一次计算。两个分支可以合并。

```cpp
if ( (!input_grad) && (! filter_grad)) {
  return;
}
下面计算~
```

Comment (chengduoZH) [reply paddle/operators/conv_transpose_op.h]: Done

File: paddle/operators/conv_transpose_op.h (hunk 1 above)
Comment (chengduoZH) [reply paddle/operators/conv_transpose_op.h]: 编译没有warning，不过最好还是改一下

**Pull request merged**